### DISCUSSION TITLE
lora training

### DISCUSSION THREAD

**Discussion by nanaj96** [chengzeyi/stable-fast#37] [General]
can we possibly lora training in comfyui with stable-fast?

In ComfyUI it is not possible. But if you use diffusers or any UI using it, I could demonstrate it for you.

**Comment by chengzeyi** [comment]
In ComfyUI this is hard to implement. But if you use diffusers or any UI backened by it and you are really interested in it, I could demonstrate it for you.

**Reply by L4Ph** [reply]
I am interested in this, what is the current status?

**Reply by chengzeyi** [reply]
I have a script for dreambooth training, I would demonstrate it later.

It is expected to work, but the training results are not guaranteed since a lot of optimizations in `stable-fast` might cause numeric issues.

**Reply by L4Ph** [reply]
Really!
I hope they merge it into [sd-script](https://github.com/kohya-ss/sd-scripts) one of these days!

You could really start a revolution when you get there!

**Reply by chengzeyi** [reply]
I have just done some experiments. Sadly speedup is not significant. 🤔If it needs to be effective, I  think I need to build a new compiler frontend with `torch.fx` and `aot_autograd`.

Current frontend with `torch.jit` lacks the ability to optimize backward graph and is incompatible with many features of `accelerate`, the widely used training acceleration framework. With `torch.jit`, we cannot explicitly capture the backward graph and wrapping the backward pass of an arbitrary model into a `CUDA graph` is also non-trivial🤔.

Currently the test results show that the benifit of `stable-fast` over `torch.compile` in `LoRA training` is small or even worse.

**Reply by chengzeyi** [reply]
And for training fusion is limited since a lot of intermediate buffers have to be generated and saved for calculating gradient during backward pass, so the speedup won't be as significant as pure inference🤔.

**Reply by chengzeyi** [reply]
Ok, after upgrading torch and refactoring stable-fast, it now works with aot_autograd! But, it is not optimized yet so the speed is not fast! No optimized backward kernels, no optimized decompositions...

https://github.com/chengzeyi/stable-fast/blob/main/examples/optimize_train_text_to_image_lora.sh

**Reply by L4Ph** [reply]
yupyupyupyup 👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍
